### ISSUE TITLE
Implement Exponential Backoff

### ISSUE THREAD

**Issue by Skylion007** [rom1504/img2dataset#331]
I am currently using this tool on a website that uses AWS CloudFront to host all their images. However, if you do too many queries to their urls, you will receive a 429 Error that does not provide any Retry After header field. This means that img2dataset will just keep hammering the CloudFront causing the ban to be extended. The only way to still allow images to be downloaded is to have an exponential backoff applied. This is also just good practice when scraping websites so it would be great if it could be included.

**Comment by Skylion007** [comment]
Yeah, unfortunately, it is unclear what the limit is. The docs say we
should implement Exponential Backoff which is probably a good policy
uniformly.

**Comment by rom1504** [comment]
it does not make sense to implement exponential backoff while maximizing the number of parallel calls.
Simply reduce the number of parallel calls

**Comment by Skylion007** [comment]
@rom1504 Depending on the settings of the host though, you could hit the rate limit with non-parallel calls in a extreme case though.

**Comment by rom1504** [comment]
yes, but in that case maybe just do a loop with a wget and a sleep and it'll be simpler and not slower

**Comment by ldfandian** [comment]
> I am currently using this tool on a website that uses AWS CloudFront to host all their images. However, if you do too many queries to their urls, you will receive a 429 Error that does not provide any Retry After header field. This means that img2dataset will just keep hammering the CloudFront causing the ban to be extended. The only way to still allow images to be downloaded is to have an exponential backoff applied. This is also just good practice when scraping websites so it would be great if it could be included.

It probably means they don't want you to crawl it~

**Comment by Skylion007** [comment]
Not necessarily, they just don't want you crawling it that fast. Some websites even send a retry after header to tell you to slowdown and try after x seconds.
